src/mainwindow/mainwindow.cpp

```diff
@@ -1133,20 +1166,35 @@ bool MainWindow::saveFile(const QString &fileName, bool baseMapOnly )
   progressDlg->setValue(0);
   progressDlg->show();
 
-  std::auto_ptr<AbstractMapStorage> storage( new MapStorage(*m_mapData , fileName, &saver.file()) );
+  std::auto_ptr<AbstractMapStorage> storage;
+  if ( mode == SAVE_WEB )
+    storage.reset( new JsonMapStorage(*m_mapData , fileName ) );
+  else
+    storage.reset( new MapStorage(*m_mapData , fileName, &saver.file()) );
+
   connect(storage->progressCounter(), SIGNAL(onPercentageChanged(quint32)), this, SLOT(percentageChanged(quint32)));
   connect(storage.get(), SIGNAL(log(const QString&, const QString&)), this, SLOT(log(const QString&, const QString&)));
 
   disableActions(true);
   //getCurrentMapWindow()->getCanvas()->hide();
-  if (storage->canSave()) storage->saveData( baseMapOnly );
+  if (storage->canSave())
+  {
+    bool baseMapOnly = mode != SAVE_FULL;
+    if ( !storage->saveData( baseMapOnly ) )
+    {
+      getCurrentMapWindow()->getCanvas()->setEnabled(true);
+      progressDlg->hide();
+      disableActions(true);
+      return false;
```

Comment:
We aren't closing the saver here.

Reply:
Can we refactor this code to be simpler? I don't like the fact that the branching logic has semi-duplicate code regarding the UI.

Reply:
The saver does not need explicit closing, its destructor handles that.

I refactored the logic in a separate `json4` branch, coming soon to a PR close to you...